plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,144 @@
+import com.mapbox.maps.plugin.viewport.experimental.state.ViewportStateChangedObserver
+import com.mapbox.maps.plugin.viewport.experimental.transition.TransitionEndListener
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportTransitionOptions
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.CopyOnWriteArraySet
+
+/**
+ * Mapbox default implementation for [ViewportPlugin].
+ */
+class ViewportPluginImpl : ViewportPlugin {
+  private val registeredStateChangedObservers = CopyOnWriteArrayList<ViewportStateChangedObserver>()
+
+  override val states = CopyOnWriteArraySet<ViewportState>()
+  override var currentState: ViewportState? = null
+    private set
+
+  override fun addViewportState(state: ViewportState) {
+    states.add(state)
+  }
+
+  override fun removeViewportState(state: ViewportState) {
+    if (currentState == state) {
+      state.cleanUp()
+      currentState = null
+    }
+    states.remove(state)
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      state,
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_STATE_TRANSITION_MAX_DURATION_MS)
+        .apply(stateTransitionOptionsBlock).build(),
+      ViewportTransitionOptions.Builder().maxDuration(DEFAULT_FRAME_TRANSITION_MAX_DURATION_MS)
+        .apply(frameTransitionOptionsBlock).build(),
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    state: ViewportState,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    states.add(state)
+    currentState?.cleanUp()
+    state.transitionToState(
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+    currentState = state
+    notifyStateChanged()
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    frameTransitionOptionsBlock: ViewportTransitionOptions.Builder.() -> Unit,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptionsBlock,
+      frameTransitionOptionsBlock,
+      transitionEndListener
+    )
+  }
+
+  override fun transitionToState(
+    stateId: String,
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    transitionToState(
+      states.first { it.id == stateId },
+      stateTransitionOptions,
+      frameTransitionOptions,
+      transitionEndListener
+    )
+  }
```

Comment:
Not a fan of the ID-based APIs. Doesn't seem like it provides much value vs just passing the state instances.

Reply:
that means user should keep the state instances by themselves? And there seems to be less value to keep the state instances inside the viewport plugin.

My imagination is that the plugin will keep reference to all the state instances, and allow user to simply transition to the target state with its ID.

Reply:
Updated the PR to remove ID-based APIs, users would need to create the ViewportState instances and add them to the ViewportPlugin before usage.